syntax = "proto3";
package GRPCGlobalVariable;

// 기존 전역 변수 관련 메시지
message Empty {
}

message GInt {
    int32 idx = 1;
    int32 val = 2;
}

message IntVal {
    int32 val = 1;
}

message GInts {
    int32 idx = 1;
    repeated int32 val = 2;
}

message IntVals {
    repeated int32 val = 1;
}

// id와 함께 문자열을 주고받는 메시지
message StringWithId {
    int32 id = 1;
    string val = 2;
}

// id만 담는 메시지 (읽기 요청에 사용)
message StringId {
    int32 id = 1;
}

// float을 주고받는 메시지
message GFloat {
    int32 idx = 1;
    float val = 2;
}

message FloatVal {
    float val = 1;
}

message GFloats {
    int32 idx = 1;
    repeated float val = 2;
}

message FloatVals {
    repeated float val = 1;
}


service GRPCGlobalVariableTask {
    // 기존 전역 변수 관련 RPC
    rpc SetInt(GInt) returns (Empty) {}
    rpc GetInt(IntVal) returns (IntVal) {}

    rpc SetInts(GInts) returns (Empty) {}
    rpc GetInts(GInt) returns (IntVals) {}

    // id와 함께 문자열을 설정합니다.
    rpc SetStringWithId(StringWithId) returns (Empty) {}
    
    // id에 해당하는 문자열을 읽어옵니다.
    rpc GetStringWithId(StringId) returns (StringWithId) {}

    // float 관련 RPC
    rpc SetFloat(GFloat) returns (Empty) {}
    rpc GetFloat(IntVal) returns (FloatVal) {}
    rpc SetFloats(GFloats) returns (Empty) {}
    rpc GetFloats(GInt) returns (FloatVals) {}
}
